### PR TITLE
BUG: DTI/TDI/PI get_indexer_non_unique with incompatible dtype

### PR DESCRIPTION
**Pull request by jbrockmendel**
I'm pretty sure we can share more code here, but PeriodEngine needs some work first.

### CODE REVIEW

**Comment by jreback** [comment]
lgtm. needs a rebase. is this user visible at a high level? e.g. in .loc or someting?

**Comment by jbrockmendel** [comment]
>  is this user visible at a high level? e.g. in .loc or someting?

Yes, will update the test introduced in #30446

**Comment by jreback** [comment]
also needs a rebase

**Comment by jbrockmendel** [comment]
hmm this should probably raise a KeyError instead of a TypeError?

```
df = pd.DataFrame(
    np.random.randn(5, 3),
    columns=["a", "b", "c"],
    index=pd.date_range("2012", freq="H", periods=5),
)

df = df.iloc[[0, 2, 2, 3]].copy() 

dti = df.index
tdi = pd.TimedeltaIndex(dti.asi8)  # matching i8 values

with pytest.raises(TypeError, match="is not convertible to datetime"):
    df.loc[tdi]

with pytest.raises(TypeError, match="is not convertible to datetime"):
    df["a"].loc[tdi]
```

**Comment by jreback** [comment]
> hmm this should probably raise a KeyError instead of a TypeError?
> 
> ```
> df = pd.DataFrame(
>     np.random.randn(5, 3),
>     columns=["a", "b", "c"],
>     index=pd.date_range("2012", freq="H", periods=5),
> )
> 
> df = df.iloc[[0, 2, 2, 3]].copy() 
> 
> dti = df.index
> tdi = pd.TimedeltaIndex(dti.asi8)  # matching i8 values
> 
> with pytest.raises(TypeError, match="is not convertible to datetime"):
>     df.loc[tdi]
> 
> with pytest.raises(TypeError, match="is not convertible to datetime"):
>     df["a"].loc[tdi]
> ```

hmm, yep sounds right

File: pandas/tests/indexing/test_loc.py [hunk 1]
```diff
@@ -1073,3 +1073,25 @@ def test_loc_slice_disallows_positional():
...
+    df = df.iloc[[0, 2, 2, 3]].copy()
+
+    dti = df.index
+    tdi = pd.TimedeltaIndex(dti.asi8)  # matching i8 values
+
+    msg = r"None of \[TimedeltaIndex.* are in the \[index\]"
+    with pytest.raises(KeyError, match=msg):
+        df.loc[tdi]
+
+    with pytest.raises(KeyError, match=msg):
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: i would test the reverse as well (e.g. TDI is the index with a DTI as indexer). This should also raise for Period, I think? (these can be handled as followons)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**